src/github.com/mozilla-services/pushgo/simplepush/structs.go

```diff
@@ -29,48 +28,32 @@ type PushCommand struct {
 }
 
 type PushWS struct {
-	uaidLock sync.RWMutex
-	uaid     string // Hex-encoded client ID; not normalized
-	Socket   Socket // Remote connection
+	uaid   string // Hex-encoded client ID; not normalized
+	Socket Socket // Remote connection
 	Store
 	Logger    *SimpleLogger
 	Metrics   *Metrics
 	Born      time.Time
-	closeLock sync.RWMutex
-	closed    bool
+	closeOnce Once
 }
 
 func (ws *PushWS) UAID() (uaid string) {
-	ws.uaidLock.RLock()
-	uaid = ws.uaid
-	ws.uaidLock.RUnlock()
-	return
+	return ws.uaid
 }
 
 func (ws *PushWS) SetUAID(uaid string) {
-	ws.uaidLock.Lock()
 	ws.uaid = uaid
-	ws.uaidLock.Unlock()
 }
 
 func (ws *PushWS) IsClosed() (closed bool) {
```

Comment:
The reason behind `IsClosed()` is convoluted...short version: it mitigates a race condition caused by reconnecting clients. Here's the long version; maybe it'll make it easier to come up with a proper solution:

The issue crops up when a client drops its connection ("connection A"), then reconnects with the same ID ("connection B"). A and B are handled by separate goroutines, each with an active [`Run()` loop](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/handlers_socket.go#L138). When B connects, we assume A is stale and should be replaced. So we [close A](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/worker.go#L479-485), which unblocks its `Run()` loop and [calls `Bye()`](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/handlers_socket.go#L131) in a `defer` block. This cleanup also runs if the client gracefully closes its connection.

Because the cleanup happens in A's goroutine, B's goroutine doesn't know when `Bye()` has returned—it only closes A, then adds B to the client map. Since A and B share the same device ID, there's a race where B is added to the map before goroutine A cleans up. In that case, when goroutine A calls `Bye()`, it'll remove the newly-added B from the map.

The `IsClosed()` hack ensures we only remove a client from the map once. When goroutine B closes stale connection A, it removes A from the map and sets `IsClosed()` to true. When goroutine A then goes to clean up, it'll see the connection has already been closed and do nothing.

Does that make sense? Is there a better way out?